### PR TITLE
Improve hooks in nuxt.config.js

### PR DESCRIPTION
**Pull request by galvez**
Related docs PR: https://github.com/nuxt/docs/pull/716

Currently they're set as follows:

```
hooks: {
  'build:done' (ctx) { }
```

With this change, they will **also** be able to set using hierarchical syntax:

```
hooks: {
  build: {
    done(ctx){ }
  }
}
```

I've already submitted a PR do nuxt/docs with a new section about it.

### CODE REVIEW

**Comment by Atinux** [comment]
Good to me, could you please make a PR to nuxt/docs and update the release note as well @galvez ?

File: lib/core/nuxt.js [hunk 1]
```diff
@@ -105,11 +105,30 @@ export default class Nuxt {
...
+      } else {
+        const value = hooksObj[key]
+        key = `${stack}:${key}`
+        keyList.push({
+          name: key.slice(1),
+          hook: value
+        })
+      }
+    })
+    return keyList.reduce((hash, pair) => {
```
Comment (clarkdo) [inline lib/core/nuxt.js]: Convert an array to an object here, and then loop the object as an array in `addObjectHooks`, does it intend to ? How about use `keysList` as an object or just return array ?

File: test/unit/basic.dev.test.js [hunk 2]
```diff
@@ -26,11 +27,16 @@ describe('basic dev', () => {
       }
     })
     nuxt = new Nuxt(config)
-    await new Builder(nuxt).build()
+    builder = new Builder(nuxt)
```
Comment (clarkdo) [inline test/unit/basic.dev.test.js]: It's fine here, but acutally `build()` also returns `builder` object.

Comment (galvez) [reply test/unit/basic.dev.test.js]: @clarkdo gotcha.

File: lib/core/nuxt.js (hunk 1 above)
Comment (galvez) [reply lib/core/nuxt.js]: >  How about use keysList as an object or just return array ?

I would probably have to rework the recursion -- not sure. I think it's mostly fine as-is.

**Review by clarkdo — APPROVED** [review]
OK, I'll refactor it later

**Pull request merged**